ReactiveSwift.playground/Pages/Sandbox.xcplaygroundpage/Contents.swift

```diff
@@ -22,4 +22,186 @@ import Foundation
  A place where you can build your sand castles 🏖.
 */
 
+enum PromiseResult<Value, Error: Swift.Error> {
+    case fulfilled(Value)
+    case rejected(Error?)   // nil error means that the producer was interrupted or empty
+    
+    init(value: Value) {
+        self = .fulfilled(value)
+    }
+    
+    init(error: Error?) {
+        self = .rejected(error)
+    }
+    
+    var value: Value? {
+        switch self {
+        case .fulfilled(let value): return value
+        case .rejected(_): return nil
+        }
+    }
+    
+    var error: Error? {
+        switch self {
+        case .fulfilled(_): return nil
+        case .rejected(let error): return error
+        }
+    }
+}
+
+protocol Thenable {
+    associatedtype Value
+    associatedtype Error: Swift.Error
+    
+    func await(notify: @escaping (PromiseResult<Value, Error>) -> Void) -> Disposable
+}
+
+extension Thenable {
+    
+    @discardableResult
+    func chain<T: Thenable>(on: Scheduler = ImmediateScheduler(), body: @escaping (PromiseResult<Value, Error>) -> T) -> Promise<T.Value, Error> where T.Error == Error {
+        return Promise<T.Value, Error> { resolve in
+            return self.await { result in
+                on.schedule {
+                    let thenable = body(result)
+                    thenable.await(notify: resolve)
+                }
+            }
+        }
+    }
+    
+    @discardableResult
+    func then<T: Thenable>(on: Scheduler = ImmediateScheduler(), body: @escaping (Value) -> T) -> Promise<T.Value, Error> where T.Error == Error {
+        
+        return self.chain { (result) -> T in
+            guard let value = result.value else {
+                return Promise<T.Value, Error>(error: result.error) as! T
+            }
+            return body(value)
+        }
+        
+    }
+}
+
+struct Promise<Value, Error: Swift.Error>: Thenable, SignalProducerConvertible {
+    typealias Result = PromiseResult<Value, Error>
+    
+    enum State {
+        case pending
+        case resolved(Result)
+        
+        var promiseResult: Result? {
+            switch self {
+            case .pending: return nil
+            case .resolved(let result): return result
+            }
+        }
+    }
+    
+    private let state: Property<State>
+    
+    private let (_lifetime, _lifetimeToken) = Lifetime.make()
+    
+    var lifetime: Lifetime { return _lifetime }
+    
+    var producer: SignalProducer<Value, Error> {
```

Comment:
I think this should be `SignalProducer.init(_: Promise<Value, Error>)` instead.

Reply:
Yeah, agreed. `state` can be made `fileprivate` and an extension for `SignalProducer` could be defined. Also it would be cleaner to make this a `let` property cause there's no reason in creating a new signal producer for every caller

Reply:
done